src/Castle.ts

```diff
@@ -0,0 +1,218 @@
+import fetch from 'node-fetch';
+import { reduce } from 'lodash';
+import { IncomingHttpHeaders } from 'http2';
+import AbortController from 'abort-controller';
+
+const defaultApiUrl = 'https://api.castle.io';
+
+type TrackParameters = {
+  event: string;
+  user_id: string;
+  user_traits: any;
```

Comment:
technically traits can be either a hash or a nil, can't we somehow check that?

Reply:
If you mean verify that the passed in thing is hash or nil, that's not something that typescript can do. TypeScript doesn't exist at runtime. But since it can be nil, the parameter should be marked as optional, fixing.

Reply:
I know it doesn't but I would expect it to have some sort of a type validating lib for the runtime that would prevent it from sending us something completely off (like a string)

Reply:
I suggested that to @wallin but he said you removed that from the ruby lib. Ideally I'd like to validate all the inputs and provide debug info for implementers.

Reply:
@wallin did I? :D You mean `castle-ruby`?

Reply:
I suggested I could do it in a non-invasive way, using `console.warn` rather than throwing an error.

Reply:
Ah ok. Didn't understand. I'm not really sure how it should work in the node ecosystem as my knowledge towards it is limited. I will rely on you in that matter as long as we have a direct way to notify developers that they're trying to do something shady :)

Reply:
The way I see it, throwing errors can be annoying as you might break the implementers code. Providing some kind of debug info is useful, as the feedback loop is tighter, but at least now we have the debugging console in the dashboard.

ATM the only two faulty inputs that cause errors in the node SDK is missing apiSecret when constructing the Castle object, and missing event when calling authenticate/track. Those are probably the only two we want to fail. So the next step would be to give feedback to implementer when they are implementing, without breaking production code. In that case, I'd prefer warnings over errors. Alternatively we can have `debug_mode` that turns on input validation.

Reply:
Input validation is not pretty in JS... Ideally, I could use a schema validator, like https://github.com/hapijs/joi. But then I have to write the type definitions twice, once for TS and once for the library.

Otherwise it's a whole bunch of these `if (params.user_traits && typeof params.user_traits !== 'object') {`

Reply:
I would be in favour of aggresive `debug_mode` or `development_mode` (whatever is standard for node) but as I said, I'm by no means an expert in node so I will rely on your expertise.